core/services/ocr2/plugins/ocr2keeper/evm21/upkeep_state_store.go

```diff
@@ -0,0 +1,175 @@
+package evm
+
+import (
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"sync"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+
+	mapset "github.com/deckarep/golang-set/v2"
+)
+
+const Separator = "|"
+
+type upkeepState struct {
+	payload  *ocr2keepers.UpkeepPayload
+	state    *UpkeepState
+	block    int64
+	upkeepId string
+}
+
+type UpkeepState uint8
+
+const Performed UpkeepState = iota
+
+type UpkeepStateReader interface {
+	// SelectByID retrieves a single upkeep state
+	SelectByID(ID string) (*ocr2keepers.UpkeepPayload, *UpkeepState, error)
+	// SelectByBlockRange retrieves upkeep states within block range from start (inclusive) to end (exclusive)
+	SelectByBlockRange(start, end int64) ([]*ocr2keepers.UpkeepPayload, []*UpkeepState, error)
+	// SelectByUpkeepIDs retrieves upkeep states for provided upkeeps
+	SelectByUpkeepIDs([]*big.Int) ([]*ocr2keepers.UpkeepPayload, []*UpkeepState, error)
+	// SelectByUpkeepIDsAndBlockRange retrieves upkeep states for provided upkeep ids and block range
+	SelectByUpkeepIDsAndBlockRange(upkeepIds []*big.Int, start, end int64) ([]*ocr2keepers.UpkeepPayload, []*UpkeepState, error)
```

Comment:
We don't really need other select functions besides `SelectByUpkeepIDsAndBlockRange`

Reply:
done